### DISCUSSION TITLE
Discrepancy in Trial Results Between Two-Step(save and resume) and One-Step Study Approaches

### DISCUSSION THREAD

**Discussion by GhsJahanmir** [optuna/optuna#5690] [Q&A]
Dear All,

I encountered an issue while testing two approaches.

I conducted two studies to compare the results. I expected the same results in both the list of hyperparameter combinations for all trials and the objective value of each trial (in my case, AUC).

In the first study, I performed 30 trials and saved the trials along with the state of the sampler at the end of the optimization using the following code.

```python
storage_url = "sqlite:///trial_27.db"             the same for three studies
study_name = "Step1_trial_27"               same for first and second study (which is the resumption of first study)
```

 The seed is fixed in the default sampler, as well as the train-test split, which I placed its code line at the beginning of the objective function code.
After objective function definition, I create study as follows and saved the state of sampler in a pickle file.

```python
study = optuna.create_study(storage=storage_url,sampler=optuna.samplers.TPESampler(seed=seed), study_name=study_name, direction='maximize', load_if_exists=True)
study.optimize(objective, n_trials=30)
with open("sampler_data_trial_27.pkl", "wb") as fout:
    pickle.dump(study.sampler, fout)
```

   
I then reloaded the sampler using the same study name and database name and continued with 30 more trials in the second study.

```python
restored_sampler = pickle.load(open("sampler_data_trial_27.pkl", "rb"))
study = optuna.create_study(storage=storage_url,sampler=restored_sampler, study_name=study_name, direction='maximize', load_if_exists=True)
study.optimize(objective, n_trials=30)
```
  

In the third study, I used the same database name but a different study name, executing the optimization for 60 trials.

What I observed is that the objective values and hyperparameter sets are the same up to trial 30 (trials 0 to 29). This means the results of the first 30 trials of the third study exactly matched those of the first study. However, there are discrepancies between the second 30 trials of the third study and the second study. Specifically, the hyperparameter sets from trials 30 to 33 are the same, while the AUC values differ. From trial 34 onward, the third and second studies began to diverge in both hyperparameter values and AUC.

I would be very grateful if anyone has any insights into why this is happening and how to resolve it. Suppose I need specific trial numbers, and due to my computational power limitations, I need to conduct the optimization in two steps.

Thanks in advance,
Ghs

**Comment by nzw0301** [comment]
Hi, could you share minimal reproducible codes with us?

**Comment by GhsJahanmir** [comment]
Sure

Here are the links of study 1 and 2.  Study 3 is the same as study 1 but with a different study name and 60 [trials.]
[study 2](https://drive.google.com/file/d/1byC5sT0UN1peD11fTQpcVrEXDV2feSnn/view?usp=drive_link)
[study 1](https://drive.google.com/file/d/1Z6-VqZ4Px_-8uQnJOGRvDo4Z_PVUY0Qq/view?usp=drive_link)

Sorry I can not upload the dataset
thank you for your help in advance
